### PR TITLE
JSON Values

### PR DESCRIPTION
**Pull request by andrew-wm-arthur**
This change moves the engine towards representing JSON values as Go values rather than strings.

I'm strongly considering a 
```
type JSONValue interface{}
```
There seems to be no precedent for this

### CODE REVIEW

**Comment by andrew-wm-arthur** [comment]
I'm strongly considering a `type JSONValue interface{}`. There seems to be no precedent for this in the codebase, but I believe it will enable better performance and correctness. The type of a field value is used to fork logic throughout the code base and without a json value type, we can't differentiate between something like `INT(11)` and `JSON(11)`. A json value type would also designate a value as a valid json document. This PR traverses the entire document to validate it on every `Convert()` call.

**Review by zachmu — COMMENTED** [review]
I think this is the right call given how we've talked about the storage engine implementation (and thinking about how to implement various JSON functions)

File: sql/arraytype.go [hunk 1]
```diff
@@ -129,9 +129,19 @@ func (t arrayType) SQL(v interface{}) (sqltypes.Value, error) {
 		return sqltypes.Value{}, err
 	}
 
-	val, err := json.Marshal(v)
-	if err != nil {
-		return sqltypes.Value{}, err
+	var val []byte
```
Comment (zachmu) [inline sql/arraytype.go:132]: Can we actually kill off this type? It's not actually supported in MySQL, so the semantics of it are very unclear. And it's getting all muddled in the with the JSON array logic. I'd rather have exactly one well supported implementation of arrays (use JSON).

Also kill off the EXPLODE function at the same time.

Could be a follow up after this PR to keep things smaller.

File: sql/expression/function/arraylength.go [hunk 2]
```diff
@@ -57,19 +57,27 @@ func (f *ArrayLength) WithChildren(children ...sql.Expression) (sql.Expression,
...
 	child, err := f.Child.Eval(ctx, row)
 	if err != nil {
 		return nil, err
 	}
-
 	if child == nil {
 		return nil, nil
 	}
 
+	if val, ok := child.(sql.JSONValue); ok {
```
Comment (zachmu) [inline sql/expression/function/arraylength.go:73]: Same comment here. This function is not part of MySQL and has poorly defined semantics. Kill it.

File: sql/json_value.go [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+	// semantics as Type.Compare()
+	Compare(v JSONValue) (cmp int, err error)
+	// ToString marshalls a JSONValue to a valid JSON-encoded string.
+	ToString() (string, error)
+}
+
+// SearchableJSONValue is JSONValue supporting in-place access operations.
+// The query engine can utilize these optimized access methods improve performance
+// by minimizing the need to unmarshall a JSONValue into a JSONDocument.
+type SearchableJSONValue interface {
```
Comment (zachmu) [inline sql/json_value.go:38]: Should have method doc for these interface methods. Fine to sketch out if they're not finalized

**Review by zachmu — APPROVED** [review]
LGTM. All my comments are basically about killing off the non-json array type

**Pull request merged**